src/sparseml/modifiers/smoothquant/pytorch.py

```diff
@@ -0,0 +1,269 @@
+        weight is ".*re:.*q_proj" and the activation is "re:.*self_attn_layer_norm" we
+        would match model.layer.0.p_proj to model.layer.0.self_attn_layer_norm and
+        repeat for model.layer.1 and so on
+        """
+        resolved_mappings = []
+        for to_balance, to_smooth in self.mappings:
+            to_smooth_layers = model.get_layers(to_smooth)
+            for layer_name, smooth_layer in to_smooth_layers.items():
+                if layer_name not in self.ignore:
+                    balance_layers = []
+                    for balance_suffix in to_balance:
+                        # find the submodule that matches the activation layer
+                        _, balance_layer = get_matching_layer(
+                            balance_suffix, layer_name, model.model
+                        )
+                        if balance_layer:
+                            balance_layers.append(balance_layer)
+                    # each mapping can contain multiple layers to balance, but only
+                    # one layer to smooth
+                    mapping = SmoothQuantMapping(
+                        layer_name, smooth_layer, balance_layers
+                    )
+                    resolved_mappings.append(mapping)
+        return resolved_mappings
+
+    def _setup_scale_hooks(self):
+        """
+        Attach a forward hook to each activation we want to smooth. This allows us to
+        calculate the dynamic range during calibration
+        """
+
+        def create_hook_fn(layer_name):
+            def hook_fn(module, inp, out):
+                # update the per-channel min/max output values seen during calibration
+                if isinstance(out, tuple):
+                    out = out[0]
+
+                hidden_dim = out.shape[-1]
+                out = out.view(-1, hidden_dim).abs()
+                latest_mins = torch.min(out, dim=0)[0]
+                latest_maxes = torch.max(out, dim=0)[0]
+
+                if layer_name in self.scales_:
+                    self.scales_[layer_name].min_channel_vals = torch.minimum(
+                        self.scales_[layer_name].min_channel_vals, latest_mins
+                    )
+                    self.scales_[layer_name].max_channel_vals = torch.maximum(
+                        self.scales_[layer_name].max_channel_vals, latest_maxes
+                    )
+                else:
+                    self.scales_[layer_name] = SmoothQuantScale(
+                        min_channel_vals=latest_mins, max_channel_vals=latest_maxes
+                    )
+
+            return hook_fn
+
+        for mapping in self.resolved_mappings_:
+            name = mapping.smooth_name
+            layer = mapping.smooth_layer
+            self.hooks_.append(layer.register_forward_hook(create_hook_fn(name)))
+
+    @torch.no_grad()
+    def _calibrate(self, model: ModifiableModelPyTorch, calibration_dataloader: List):
```

Comment:
probably can move this out to a helper function, we do the same in base quantization modifier right?